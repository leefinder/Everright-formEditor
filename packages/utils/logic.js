import _ from 'lodash-es'
export const checkIdExistInLogic = (fieldId, logic) => {
  return new RegExp(`"${fieldId}"`).test(JSON.stringify(logic))
}
export const removeLogicDataByid = (fieldId, logic) => {
  for (const rules of Object.values(logic)) {
    for (let i0 = 0; i0 < rules.length; i0++) {
      const conditions = _.get(rules[i0], 'ifRules.filters[0].conditions', [])
      for (let i1 = 0; i1 < conditions.length; i1++) {
        if (conditions[i1].property === fieldId) {
          conditions.splice(i1--, 1)
        }
      }
      if (!conditions.length) {
        rules.splice(i0--, 1)
      }
    }
  }
  for (const key in logic) {
    if (!logic[key].length) {
      delete logic[key]
    }
  }

  for (const rules of Object.values(logic)) {
    for (let i0 = 0; i0 < rules.length; i0++) {
      const conditions = _.get(rules[i0], 'thenRules.filters[0].conditions', [])
      for (let i1 = 0; i1 < conditions.length; i1++) {
        for (let i2 = 0; i2 < conditions[i1].value.length; i2++) {
          if (conditions[i1].value[i2] === fieldId) {
            conditions[i1].value.splice(i2--, 1)
          }
        }
        if (!conditions[i1].value.length) {
          conditions.splice(i1--, 1)
        }
      }
      if (!conditions.length) {
        rules.splice(i0--, 1)
      }
    }
  }
  for (const key in logic) {
    if (!logic[key].length) {
      delete logic[key]
    }
  }
}
