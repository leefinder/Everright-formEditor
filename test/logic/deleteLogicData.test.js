import { describe, assert, expect, test, beforeEach } from 'vitest'
import { removeLogicDataByid, checkIdExistInLogic } from '@ER/utils/logic.js'
import { dels } from '@ER-test/data/logic.js'
import _ from 'lodash-es'
describe('Delete logic data based by field id', () => {
  let testData = {}
  beforeEach(() => {
    testData = _.cloneDeep(dels)
  })
  test('Delete the data inside the if and then statement(1)', () => {
    expect(checkIdExistInLogic('6FobSGWf9gFBgK0MS7a3E', testData[0])).toBeTruthy()
    removeLogicDataByid('6FobSGWf9gFBgK0MS7a3E', testData[0])
    expect(testData[0]).toMatchSnapshot()
  })
  test('Delete the data inside the if and then statement(2)', () => {
    expect(checkIdExistInLogic('NLSfG6qSkyzXWfL82rW0E', testData[0])).toBeTruthy()
    removeLogicDataByid('NLSfG6qSkyzXWfL82rW0E', testData[0])
    expect(testData[0]).toMatchSnapshot()
  })
  test('Delete the data inside the if and then statement(3)', () => {
    expect(checkIdExistInLogic('315WAA5x2C6VvvL7kM1pF', testData[0])).toBeTruthy()
    removeLogicDataByid('315WAA5x2C6VvvL7kM1pF', testData[0])
    expect(testData[0]).toMatchSnapshot()
  })
  test('Delete the data inside the if and then statement(4)', () => {
    expect(checkIdExistInLogic('3EnU8Uc7e7gjyt0I_wkAl', testData[0])).toBeTruthy()
    removeLogicDataByid('3EnU8Uc7e7gjyt0I_wkAl', testData[0])
    expect(testData[0]).toMatchSnapshot()
  })
})
