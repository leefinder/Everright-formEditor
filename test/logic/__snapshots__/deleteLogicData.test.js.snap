// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Delete logic data based by field id > Delete the data inside the if and then statement(1) 1`] = `
{
  "required": [
    {
      "ifRules": {
        "filters": [
          {
            "conditions": [
              {
                "dateOperator": "date",
                "operator": "equal",
                "property": "NLSfG6qSkyzXWfL82rW0E",
                "value": {
                  "dateType": "static",
                  "value": [
                    "1685116800",
                  ],
                },
              },
            ],
            "logicalOperator": "and",
          },
        ],
        "logicalOperator": "and",
      },
      "thenRules": {
        "filters": [
          {
            "conditions": [
              {
                "operator": "required",
                "property": "required",
                "value": [
                  "315WAA5x2C6VvvL7kM1pF",
                ],
              },
            ],
            "logicalOperator": "and",
          },
        ],
        "logicalOperator": "and",
      },
    },
    {
      "ifRules": {
        "filters": [
          {
            "conditions": [
              {
                "dateOperator": "date",
                "operator": "equal",
                "property": "NLSfG6qSkyzXWfL82rW0E",
                "value": {
                  "dateType": "static",
                  "value": [
                    "1683043200",
                  ],
                },
              },
            ],
            "logicalOperator": "and",
          },
        ],
        "logicalOperator": "and",
      },
      "thenRules": {
        "filters": [
          {
            "conditions": [
              {
                "operator": "required",
                "property": "required",
                "value": [
                  "315WAA5x2C6VvvL7kM1pF",
                ],
              },
            ],
            "logicalOperator": "and",
          },
        ],
        "logicalOperator": "and",
      },
    },
  ],
}
`;

exports[`Delete logic data based by field id > Delete the data inside the if and then statement(2) 1`] = `
{
  "readOnly": [
    {
      "ifRules": {
        "filters": [
          {
            "conditions": [
              {
                "dateOperator": "date",
                "operator": "equal",
                "property": "6FobSGWf9gFBgK0MS7a3E",
                "value": {
                  "dateType": "static",
                  "value": "1682956800",
                },
              },
              {
                "dateOperator": "date",
                "operator": "equal",
                "property": "6FobSGWf9gFBgK0MS7a3E",
                "value": {
                  "dateType": "static",
                  "value": "1684512000",
                },
              },
            ],
            "logicalOperator": "and",
          },
        ],
        "logicalOperator": "and",
      },
      "thenRules": {
        "filters": [
          {
            "conditions": [
              {
                "operator": "readOnly",
                "property": "readOnly",
                "value": [
                  "3EnU8Uc7e7gjyt0I_wkAl",
                  "315WAA5x2C6VvvL7kM1pF",
                ],
              },
              {
                "operator": "readOnly",
                "property": "readOnly",
                "value": [
                  "3EnU8Uc7e7gjyt0I_wkAl",
                ],
              },
            ],
            "logicalOperator": "and",
          },
        ],
        "logicalOperator": "and",
      },
    },
  ],
  "required": [
    {
      "ifRules": {
        "filters": [
          {
            "conditions": [
              {
                "dateOperator": "date",
                "operator": "equal",
                "property": "6FobSGWf9gFBgK0MS7a3E",
                "value": {
                  "dateType": "static",
                  "value": "1683648000",
                },
              },
            ],
            "logicalOperator": "and",
          },
        ],
        "logicalOperator": "and",
      },
      "thenRules": {
        "filters": [
          {
            "conditions": [
              {
                "operator": "required",
                "property": "required",
                "value": [
                  "315WAA5x2C6VvvL7kM1pF",
                ],
              },
            ],
            "logicalOperator": "and",
          },
        ],
        "logicalOperator": "and",
      },
    },
    {
      "ifRules": {
        "filters": [
          {
            "conditions": [
              {
                "dateOperator": "date",
                "operator": "equal",
                "property": "6FobSGWf9gFBgK0MS7a3E",
                "value": {
                  "dateType": "static",
                  "value": "1683907200",
                },
              },
            ],
            "logicalOperator": "and",
          },
        ],
        "logicalOperator": "and",
      },
      "thenRules": {
        "filters": [
          {
            "conditions": [
              {
                "operator": "required",
                "property": "required",
                "value": [
                  "315WAA5x2C6VvvL7kM1pF",
                ],
              },
            ],
            "logicalOperator": "and",
          },
        ],
        "logicalOperator": "and",
      },
    },
  ],
  "visible": [
    {
      "ifRules": {
        "filters": [
          {
            "conditions": [
              {
                "dateOperator": "date",
                "operator": "equal",
                "property": "6FobSGWf9gFBgK0MS7a3E",
                "value": {
                  "dateType": "static",
                  "value": "1683129600",
                },
              },
            ],
            "logicalOperator": "and",
          },
        ],
        "logicalOperator": "and",
      },
      "thenRules": {
        "filters": [
          {
            "conditions": [
              {
                "operator": "field",
                "property": "show",
                "value": [
                  "315WAA5x2C6VvvL7kM1pF",
                ],
              },
              {
                "operator": "field",
                "property": "show",
                "value": [
                  "315WAA5x2C6VvvL7kM1pF",
                ],
              },
            ],
            "logicalOperator": "and",
          },
        ],
        "logicalOperator": "and",
      },
    },
  ],
}
`;

exports[`Delete logic data based by field id > Delete the data inside the if and then statement(3) 1`] = `
{
  "readOnly": [
    {
      "ifRules": {
        "filters": [
          {
            "conditions": [
              {
                "dateOperator": "date",
                "operator": "equal",
                "property": "6FobSGWf9gFBgK0MS7a3E",
                "value": {
                  "dateType": "static",
                  "value": "1682956800",
                },
              },
              {
                "dateOperator": "date",
                "operator": "equal",
                "property": "6FobSGWf9gFBgK0MS7a3E",
                "value": {
                  "dateType": "static",
                  "value": "1684512000",
                },
              },
            ],
            "logicalOperator": "and",
          },
        ],
        "logicalOperator": "and",
      },
      "thenRules": {
        "filters": [
          {
            "conditions": [
              {
                "operator": "readOnly",
                "property": "readOnly",
                "value": [
                  "NLSfG6qSkyzXWfL82rW0E",
                ],
              },
              {
                "operator": "readOnly",
                "property": "readOnly",
                "value": [
                  "3EnU8Uc7e7gjyt0I_wkAl",
                ],
              },
              {
                "operator": "readOnly",
                "property": "readOnly",
                "value": [
                  "3EnU8Uc7e7gjyt0I_wkAl",
                ],
              },
            ],
            "logicalOperator": "and",
          },
        ],
        "logicalOperator": "and",
      },
    },
  ],
  "visible": [
    {
      "ifRules": {
        "filters": [
          {
            "conditions": [
              {
                "dateOperator": "date",
                "operator": "equal",
                "property": "6FobSGWf9gFBgK0MS7a3E",
                "value": {
                  "dateType": "static",
                  "value": "1682870400",
                },
              },
            ],
            "logicalOperator": "and",
          },
        ],
        "logicalOperator": "and",
      },
      "thenRules": {
        "filters": [
          {
            "conditions": [
              {
                "operator": "field",
                "property": "show",
                "value": [
                  "NLSfG6qSkyzXWfL82rW0E",
                ],
              },
            ],
            "logicalOperator": "and",
          },
        ],
        "logicalOperator": "and",
      },
    },
  ],
}
`;

exports[`Delete logic data based by field id > Delete the data inside the if and then statement(4) 1`] = `
{
  "readOnly": [
    {
      "ifRules": {
        "filters": [
          {
            "conditions": [
              {
                "dateOperator": "date",
                "operator": "equal",
                "property": "6FobSGWf9gFBgK0MS7a3E",
                "value": {
                  "dateType": "static",
                  "value": "1682956800",
                },
              },
              {
                "dateOperator": "date",
                "operator": "equal",
                "property": "6FobSGWf9gFBgK0MS7a3E",
                "value": {
                  "dateType": "static",
                  "value": "1684512000",
                },
              },
            ],
            "logicalOperator": "and",
          },
        ],
        "logicalOperator": "and",
      },
      "thenRules": {
        "filters": [
          {
            "conditions": [
              {
                "operator": "readOnly",
                "property": "readOnly",
                "value": [
                  "NLSfG6qSkyzXWfL82rW0E",
                ],
              },
              {
                "operator": "readOnly",
                "property": "readOnly",
                "value": [
                  "315WAA5x2C6VvvL7kM1pF",
                ],
              },
            ],
            "logicalOperator": "and",
          },
        ],
        "logicalOperator": "and",
      },
    },
  ],
  "required": [
    {
      "ifRules": {
        "filters": [
          {
            "conditions": [
              {
                "dateOperator": "date",
                "operator": "equal",
                "property": "6FobSGWf9gFBgK0MS7a3E",
                "value": {
                  "dateType": "static",
                  "value": "1683648000",
                },
              },
              {
                "dateOperator": "date",
                "operator": "equal",
                "property": "NLSfG6qSkyzXWfL82rW0E",
                "value": {
                  "dateType": "static",
                  "value": [
                    "1685116800",
                  ],
                },
              },
            ],
            "logicalOperator": "and",
          },
        ],
        "logicalOperator": "and",
      },
      "thenRules": {
        "filters": [
          {
            "conditions": [
              {
                "operator": "required",
                "property": "required",
                "value": [
                  "315WAA5x2C6VvvL7kM1pF",
                ],
              },
            ],
            "logicalOperator": "and",
          },
        ],
        "logicalOperator": "and",
      },
    },
    {
      "ifRules": {
        "filters": [
          {
            "conditions": [
              {
                "dateOperator": "date",
                "operator": "equal",
                "property": "NLSfG6qSkyzXWfL82rW0E",
                "value": {
                  "dateType": "static",
                  "value": [
                    "1683043200",
                  ],
                },
              },
              {
                "dateOperator": "date",
                "operator": "equal",
                "property": "6FobSGWf9gFBgK0MS7a3E",
                "value": {
                  "dateType": "static",
                  "value": "1683907200",
                },
              },
            ],
            "logicalOperator": "and",
          },
        ],
        "logicalOperator": "and",
      },
      "thenRules": {
        "filters": [
          {
            "conditions": [
              {
                "operator": "required",
                "property": "required",
                "value": [
                  "315WAA5x2C6VvvL7kM1pF",
                ],
              },
            ],
            "logicalOperator": "and",
          },
        ],
        "logicalOperator": "and",
      },
    },
  ],
  "visible": [
    {
      "ifRules": {
        "filters": [
          {
            "conditions": [
              {
                "dateOperator": "date",
                "operator": "equal",
                "property": "6FobSGWf9gFBgK0MS7a3E",
                "value": {
                  "dateType": "static",
                  "value": "1682870400",
                },
              },
            ],
            "logicalOperator": "and",
          },
        ],
        "logicalOperator": "and",
      },
      "thenRules": {
        "filters": [
          {
            "conditions": [
              {
                "operator": "field",
                "property": "show",
                "value": [
                  "NLSfG6qSkyzXWfL82rW0E",
                ],
              },
            ],
            "logicalOperator": "and",
          },
        ],
        "logicalOperator": "and",
      },
    },
    {
      "ifRules": {
        "filters": [
          {
            "conditions": [
              {
                "dateOperator": "date",
                "operator": "equal",
                "property": "6FobSGWf9gFBgK0MS7a3E",
                "value": {
                  "dateType": "static",
                  "value": "1683129600",
                },
              },
            ],
            "logicalOperator": "and",
          },
        ],
        "logicalOperator": "and",
      },
      "thenRules": {
        "filters": [
          {
            "conditions": [
              {
                "operator": "field",
                "property": "show",
                "value": [
                  "315WAA5x2C6VvvL7kM1pF",
                ],
              },
              {
                "operator": "field",
                "property": "show",
                "value": [
                  "315WAA5x2C6VvvL7kM1pF",
                ],
              },
            ],
            "logicalOperator": "and",
          },
        ],
        "logicalOperator": "and",
      },
    },
  ],
}
`;
